src/App.tsx

```diff
@@ -1,24 +1,240 @@
+  }, []);
+
+  const handleClearCompletedTodos = useCallback(() => {
+    todos
+      .filter(({ completed }) => completed)
+      .forEach(handleDeleteTodo);
+  }, [todos]);
+
+  const handleToggleTodo = useCallback((todo: Todo) => {
+    clearErrorMessage();
+    startProcessingTodo(todo);
+
+    todoService.toggleTodo(todo.id, todo.completed)
+      .then(replaceTodoInState)
+      .catch(() => setErrorMessage(ErrorMessage.UpdateTodo))
+      .finally(() => stopProcessingTodo(todo));
+  }, []);
+
+  const handleToggleAllTodos = useCallback(() => {
+    const hasActive = todos.some(({ completed }) => !completed);
+
+    todos.forEach(todo => {
+      if (hasActive && todo.completed) {
+        return;
+      }
+
+      handleToggleTodo(todo);
+    });
+  }, [todos]);
+
+  const handleEditTodo = useCallback((todo: Todo, title: string) => {
+    clearErrorMessage();
+    startProcessingTodo(todo);
+
+    if (!title) {
+      return todoService.deleteTodo(todo.id)
+        .then(() => removeTodoFromState(todo))
+        .catch(() => {
+          setErrorMessage(ErrorMessage.DeleteTodo);
+
+          throw new Error(ErrorMessage.DeleteTodo);
+        })
+        .finally(() => stopProcessingTodo(todo));
+    }
+
+    return todoService.editTodo(todo.id, title)
+      .then(replaceTodoInState)
+      .catch(() => {
+        setErrorMessage(ErrorMessage.UpdateTodo);
+
+        throw new Error(ErrorMessage.UpdateTodo);
+      })
+      .finally(() => stopProcessingTodo(todo));
+  }, []);
+
+  const hasTodos = todos.length > 0 || !!tempTodo;
+  const hasCompletedTodos = todos.some(({ completed }) => completed);
+  const isAllCompleted = todos.every(({ completed }) => completed);
+  const activeTodosCount = todos.filter(({ completed }) => !completed).length;
+  const totalTodoCount = todos.length;
+
+  const filteredTodos = useMemo(() => todos.filter(({ completed }) => {
+    switch (filterValue) {
+      case Status.Active:
+        return !completed;
+
+      case Status.Completed:
+        return completed;
+
+      case Status.All:
+      default:
+        return true;
+    }
+  }), [todos, filterValue]);
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">React Todo App - Add and Delete</a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <Header>
+          {hasTodos && (
+            <ToggleAllButton
+              active={isAllCompleted}
+              onToggleAll={handleToggleAllTodos}
+            />
+          )}
+
+          <NewTodo
+            refocus={totalTodoCount}
+            onAdd={handleAddTodo}
+            onError={setErrorMessage}
+          />
```

Comment:
Why not to move this inside `Header` component?

Reply:
It is inside `Header` component. `Header` was implemented as a simple wrapper because it doesn't need any logic on its own and I don't want to pass a bunch of props to child components through it.

Reply:
@desevoker ok, if you don't want to expose it as a separate component to avoid doing a lot of code, then why expose as a wrapper one tag that can be used directly?
now this is an unnecessary increase in code and dependencies, and it is better then to use ordinary tags

Reply:
Ok I get it. But it is still a valid solution, just a different one.